src/core/experimental/array.d

```diff
@@ -0,0 +1,2299 @@
+            convFormat("Cannot emplace a %1$s because %1$s.this(this) is annotated with @disable.", T.stringof));
+
+        //We can't emplace.
+        static assert(false,
+            convFormat("%s cannot be emplaced from %s.", T.stringof, Args[].stringof));
+    }
+}
+// ditto
+package void emplaceRef(UT, Args...)(ref UT chunk, auto ref Args args)
+if (is(UT == Unqual!UT))
+{
+    emplaceRef!(UT, UT)(chunk, args);
+}
+
+//emplace helper functions
+private void emplaceInitializer(T)(scope ref T chunk) @trusted pure nothrow
+{
+    static if (__traits(isZeroInit, T))
+    {
+        import core.stdc.string : memset;
+        memset(&chunk, 0, T.sizeof);
+    }
+    else
+    {
+        import core.stdc.string : memcpy;
+        static immutable T init = T.init;
+        memcpy(&chunk, &init, T.sizeof);
+    }
+}
+
+private @nogc pure nothrow @safe
+void testEmplaceChunk(void[] chunk, size_t typeSize, size_t typeAlignment)
+{
+    assert(chunk.length >= typeSize, "emplace: Chunk size too small.");
+    assert((cast(size_t) chunk.ptr) % typeAlignment == 0, "emplace: Chunk is not aligned.");
+}
+
+enum hasElaborateDestructor(T) = __traits(compiles, { T t; t.__dtor(); })
+                                 ||  __traits(compiles, { T t; t.__xdtor(); });
+
+void dispose(A, T)(auto ref A alloc, auto ref T* p)
+{
+    static if (hasElaborateDestructor!T)
+    {
+        destroy(*p);
+    }
+    alloc.deallocate((cast(void*) p)[0 .. T.sizeof]);
+    static if (__traits(isRef, p))
+        p = null;
+}
+
+void dispose(A, T)(auto ref A alloc, auto ref T p)
+if (is(T == class) || is(T == interface))
+{
+    if (!p) return;
+    static if (is(T == interface))
+    {
+        version (Windows)
+        {
+            import core.sys.windows.unknwn : IUnknown;
+            static assert(!is(T: IUnknown), "COM interfaces can't be destroyed in "
+                ~ __PRETTY_FUNCTION__);
+        }
+        auto ob = cast(Object) p;
+    }
+    else
+        alias ob = p;
+    auto support = (cast(void*) ob)[0 .. typeid(ob).initializer.length];
+    destroy(p);
+    alloc.deallocate(support);
+    static if (__traits(isRef, p))
+        p = null;
+}
+
+void dispose(A, T)(auto ref A alloc, auto ref T[] array)
+{
+    static if (hasElaborateDestructor!(typeof(array[0])))
+    {
+        foreach (ref e; array)
+        {
+            destroy(e);
+        }
+    }
+    alloc.deallocate(array);
+    static if (__traits(isRef, array))
+        array = null;
+}
+
+// } Allocators
+
+// } End "Imports" from Phobos
+
+/**
+The element type of `R`. `R` does not have to be a range. The element type is
+determined as the type yielded by `r[0]` for an object `r` of type `R`.
+ */
+template ElementType(R)
+{
+    static if (is(typeof(R.init[0].init) T))
+        alias ElementType = T;
+    else
+        alias ElementType = void;
+}
+
+auto tail(Collection)(Collection collection)
+if (is(typeof(collection.popFront())))
```

Comment:
This was rejected for Phobos:

https://github.com/dlang/phobos/pull/4010

I would still love to see it happen though.

Reply:
Let's make it `private` for now until we can convince Andrei to add this to `std.range`

Reply:
get rid or move